packages/ai-openai/src/node/openai-language-model.ts

```diff
@@ -253,4 +239,92 @@ export class OpenAiModel implements LanguageModel {
             return new OpenAI({ apiKey: apiKey ?? 'no-key', baseURL: this.url });
         }
     }
+
+    protected processMessages(messages: LanguageModelRequestMessage[]): ChatCompletionMessageParam[] {
+        return this.openAiModelUtils.processMessages(messages, this.developerMessageSettings, this.model);
+    }
+}
+
+/**
+ * Utility class for processing messages for the OpenAI language model.
+ *
+ * Adopters can rebind this class to implement custom message processing behavior.
+ */
+@injectable()
+export class OpenAiModelUtils {
+
+    protected processSystemMessages(
+        messages: LanguageModelRequestMessage[],
+        developerMessageSettings: DeveloperMessageSettings
+    ): LanguageModelRequestMessage[] {
+        if (messages.length > 0 && messages[0].actor === 'system') {
```

Comment:
I'm wondering whether we should really limit this behavior for system messages that are on the first position. Shouldn't we just iterate through the list and apply this logic to all system messages wherever they are?

We could then maybe rename `mergeWithFirstUserMessage` with `mergeWithNextUserMessage`.

Reply:
Agreed. The interesting cases are then:

- system:1
- system:2
- assistant:1

I would merge this to: 
- user:1\n2
- assistant:1

and 
- system:1
- system:2
- user:A

i would merge this to
- user: 1\n2\nA

Reasoning would be that there are APIs which forbid two following user messages.
We can easily achieve this behavior by running from last to first while merging.

@planger

Reply:
Yes, this looks good and makes sense! Thank you!

Reply:
I change the value to "mergeWithFollowingUserMessage". It only merges, if the next message is a user message. Otherwise, I fear we risk very strange changes in the order of message. You can of course not produce message orders in which this might break specific models, e.g.

- user
- system
- assistant
- user

will become
- user
- user
- assistant
- user

But if you have these super special cases, you can now still override the processMessages.